File: ios/Keyboard/UI/KeyButton.swift
import UIKit

/// A single key. A `UIControl` so we get precise touch phases for: press
/// highlight, slide-off cancel, a preview bubble on character/punctuation
/// keys, press-and-hold repeat for backspace, and press-and-hold alternates
/// on grouped punctuation (hold expands the bubble into a row; sliding moves
/// the selection; release commits it).
final class KeyButton: UIControl {
    let spec: KeySpec
    var onTap: ((KeyCap) -> Void)?
    /// Non-clipping view the preview bubble is added to (so it can draw above keys).
    weak var popupHost: UIView?

    /// Overrides the displayed glyph (e.g. uppercase letters when shift is on,
    /// or ⇪ for caps-lock). Defaults to `spec.cap.label`.
    var displayText: String {
        didSet { titleLabel.text = displayText }
    }
    /// When true (shift armed/locked), render the key in the "active" light style.
    var isAccented = false {
        didSet { applyColors(pressed: false) }
    }

    private let titleLabel = UILabel()
    private var theme: KeyboardTheme
    private var popup: KeyPopupView?
    private var delayTimer: Timer?
    private var repeatTimer: Timer?
    private var alternatesTimer: Timer?
    /// True once a press-and-hold expanded the popup into the alternates row:
    /// tracking then slides the selection instead of press-highlighting.
    private var showingAlternates = false

    private var isRepeating: Bool { spec.cap == .backspace }

    init(spec: KeySpec, theme: KeyboardTheme) {
        self.spec = spec
        self.theme = theme
        self.displayText = spec.cap.label
        super.init(frame: .zero)

        applyChrome()

        titleLabel.text = displayText
        titleLabel.textAlignment = .center
        titleLabel.font = Self.font(for: spec.cap)
        titleLabel.adjustsFontSizeToFitWidth = true
        titleLabel.minimumScaleFactor = 0.6
        titleLabel.translatesAutoresizingMaskIntoConstraints = false
        addSubview(titleLabel)
        NSLayoutConstraint.activate([
            titleLabel.centerXAnchor.constraint(equalTo: centerXAnchor),
            titleLabel.centerYAnchor.constraint(equalTo: centerYAnchor),
            titleLabel.leadingAnchor.constraint(greaterThanOrEqualTo: leadingAnchor, constant: 2),
        ])
        applyColors(pressed: false)
    }

    required init?(coder: NSCoder) { fatalError("init(coder:) has not been implemented") }

    func apply(theme: KeyboardTheme) {
        self.theme = theme
        applyChrome()
        applyColors(pressed: false)
    }

    /// Style-dependent key shape and shadow (see `KeyboardTheme`'s chrome vars).
    private func applyChrome() {
        layer.cornerRadius = theme.keyCornerRadius
        layer.cornerCurve = theme.keyCornerCurve
        layer.shadowColor = theme.keyShadow.cgColor
        layer.shadowOpacity = theme.keyShadowOpacity
        layer.shadowRadius = theme.keyShadowBlur
        layer.shadowOffset = CGSize(width: 0, height: 1)
    }

    private static func font(for cap: KeyCap) -> UIFont {
        switch cap {
        case .char, .insertLiteral: return .systemFont(ofSize: 22)
        case .space, .ret, .toLayer, .globe, .spacer: return .systemFont(ofSize: 16)
        case .shift, .backspace: return .systemFont(ofSize: 20)
        }
    }

    // MARK: - Colors

    private var isLightKey: Bool {
        switch spec.cap {
        case .char, .insertLiteral, .space: return true
        default: return false
        }
    }

    private func applyColors(pressed: Bool) {
        titleLabel.textColor = isAccented ? theme.shiftActiveText
            : isReturnKey ? theme.returnText
            : (isLightKey ? theme.keyText : theme.specialKeyText)

        let base: UIColor
        if isAccented {
            base = theme.shiftActiveBackground  // active shift: light
        } else if isReturnKey {
            base = theme.returnBackground
        } else if isLightKey {
            base = theme.keyBackground
        } else {
            base = theme.specialKeyBackground
        }
        backgroundColor = pressed ? (isLightKey ? theme.keyHighlight : theme.keyBackground) : base
    }

    private var isReturnKey: Bool { spec.cap == .ret }

    // MARK: - Touch tracking

    override func beginTracking(_ touch: UITouch, with event: UIEvent?) -> Bool {
        KeyClick.play(spec.cap)
        applyColors(pressed: true)
        showPopupIfPreviewable()
        if !spec.alternates.isEmpty { startAlternatesDelay() }
        if isRepeating { fire(); startRepeat() }
        return true
    }

    override func continueTracking(_ touch: UITouch, with event: UIEvent?) -> Bool {
        if showingAlternates, let popup {
            popup.updateSelection(forTouch: touch.location(in: popup))
            return true
        }
        let inside = hitSlop.contains(touch.location(in: self))
        applyColors(pressed: inside)
        if !inside {
            hidePopup()
            stopRepeat()
        }
        return true
    }

    override func endTracking(_ touch: UITouch?, with event: UIEvent?) {
        applyColors(pressed: false)
        stopRepeat()
        if showingAlternates {
            let selected = popup?.selectedValue
            hidePopup()
            if let selected {
                // The primary commits like a plain tap (a future .char group
                // must still go through the engine); alternates bypass it.
                selected == displayText ? fire() : onTap?(.insertLiteral(selected))
            }
            return
        }
        hidePopup()
        let inside = touch.map { hitSlop.contains($0.location(in: self)) } ?? false
        if !isRepeating, inside { fire() }
    }

    override func cancelTracking(with event: UIEvent?) {
        applyColors(pressed: false)
        hidePopup()
        stopRepeat()
    }

    private var hitSlop: CGRect { bounds.insetBy(dx: -8, dy: -8) }

    private func fire() { onTap?(spec.cap) }

    /// QA hook: render the pressed state + preview bubble without a touch.
    /// Popups only live during a touch, so screenshots need this (see the
    /// preview app's `-popup` launch arg).
    func showPressedForQA() {
        applyColors(pressed: true)
        showPopupIfPreviewable()
    }

    /// QA hook: render the expanded press-and-hold row with the given group
    /// index highlighted (0 = primary; see the preview app's `-alts` arg).
    func showAlternatesForQA(selected: Int) {
        applyColors(pressed: true)
        expandAlternates()
        popup?.selectValue(at: selected)
    }

    // MARK: - Repeat

    private func startRepeat() {
        delayTimer = Timer.scheduledTimer(withTimeInterval: 0.35, repeats: false) { [weak self] _ in
            self?.repeatTimer = Timer.scheduledTimer(withTimeInterval: 0.1, repeats: true) { [weak self] _ in
                guard let self else { return }
                KeyClick.play(self.spec.cap)   // the system keyboard clicks every repeat tick
                self.fire()
            }
        }
    }

    private func stopRepeat() {
        delayTimer?.invalidate(); delayTimer = nil
        repeatTimer?.invalidate(); repeatTimer = nil
    }

    // MARK: - Popup

    private var isPreviewable: Bool {
        // Letters and direct-insert marks, like the system keyboard (special
        // keys show no bubble).
        if case .insertLiteral = spec.cap { return true }
        return spec.cap.isCharacter
    }

    private func showPopupIfPreviewable() {
        // iPhone only: the system keyboard shows no key previews on iPad,
        // where keys are big enough that the finger doesn't cover the glyph.
        guard traitCollection.userInterfaceIdiom != .pad else { return }
        guard isPreviewable, let host = popupHost else { return }
        let popup = KeyPopupView(text: displayText, theme: theme,
                                 keyFrame: convert(bounds, to: host),
                                 hostBounds: host.bounds)
        host.addSubview(popup)
        self.popup = popup
    }

    // MARK: - Alternates (press-and-hold)

    private func startAlternatesDelay() {
        alternatesTimer = Timer.scheduledTimer(withTimeInterval: 0.45, repeats: false) { [weak self] _ in
            self?.expandAlternates()
        }
    }

    /// Swap the single-glyph bubble for the sliding alternates row.
    private func expandAlternates() {
        guard traitCollection.userInterfaceIdiom != .pad else { return }
        guard !spec.alternates.isEmpty, let host = popupHost else { return }
        popup?.removeFromSuperview()
        let popup = KeyPopupView(alternates: [displayText] + spec.alternates, theme: theme,
                                 keyFrame: convert(bounds, to: host),
                                 hostBounds: host.bounds)
        host.addSubview(popup)
        self.popup = popup
        showingAlternates = true
    }

    private func hidePopup() {
        alternatesTimer?.invalidate(); alternatesTimer = nil
        showingAlternates = false
        popup?.removeFromSuperview()
        popup = nil
    }

    deinit {
        stopRepeat()
        alternatesTimer?.invalidate()
    }
}

File: android/app/src/main/java/com/hronro/imejd/ui/KeyButton.kt
// A single key, Material-styled: rounded surface with a touch ripple, centered
// glyph, slide-off cancel, press-and-hold repeat for backspace, and
// press-and-hold alternates on grouped punctuation (hold expands the balloon
// into a row; sliding moves the selection; release commits it). (Logic follows
// ios/Keyboard/UI/KeyButton.swift; the look is native Android.)
package com.hronro.imejd.ui

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.ColorStateList
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.graphics.drawable.Drawable
import android.graphics.drawable.GradientDrawable
import android.graphics.drawable.RippleDrawable
import android.os.Handler
import android.os.Looper
import android.util.TypedValue
import android.view.MotionEvent
import android.view.View
import androidx.core.content.ContextCompat
import com.hronro.imejd.R

@SuppressLint("ViewConstructor")
class KeyButton(
    context: Context,
    val spec: KeySpec,
    private var theme: KeyboardTheme,
) : View(context) {

    var onTap: ((KeyCap) -> Unit)? = null

    /** Preview-balloon lifecycle, emitted only for character keys (Char/InsertLiteral). */
    var onPreview: ((KeyButton, KeyPreviewEvent) -> Unit)? = null

    /** Press-and-hold alternates: expand the balloon into the group row (false
     *  when the owner shows no previews, e.g. tablets), slide the selection
     *  (key-local coordinates), and commit it (null when deselected). */
    var onAlternatesExpand: ((KeyButton) -> Boolean)? = null
    var onAlternatesSlide: ((KeyButton, Float, Float) -> Unit)? = null
    var onAlternatesCommit: ((KeyButton) -> String?)? = null

    var displayText: String = spec.cap.label
        set(value) { field = value; invalidate() }

    /** When true (shift armed/locked), render the key in the "active" light style. */
    var isAccented: Boolean = false
        set(value) { field = value; installBackground(); invalidate() }

    /** Special keys (shift/backspace/globe) render a vector icon instead of a glyph. */
    var icon: Drawable? = null
        set(value) { field = value?.mutate(); invalidate() }

    private val isRepeating get() = spec.cap is KeyCap.Backspace
    private val hasPreview get() = spec.cap is KeyCap.Char || spec.cap is KeyCap.InsertLiteral

    private val density = context.resources.displayMetrics.density
    private val corner = 8f * density
    private val expand = 8f * density

    private val textPaint = Paint(Paint.ANTI_ALIAS_FLAG or Paint.SUBPIXEL_TEXT_FLAG).apply {
        textAlign = Paint.Align.CENTER
        textSize = TypedValue.applyDimension(
            TypedValue.COMPLEX_UNIT_SP, fontSizeSp(spec.cap), context.resources.displayMetrics,
        )
    }

    private val handler = Handler(Looper.getMainLooper())
    private var repeatRunnable: Runnable? = null
    private var alternatesRunnable: Runnable? = null
    /** True once a press-and-hold expanded the balloon into the alternates row:
     *  tracking then slides the selection instead of press-highlighting. */
    private var showingAlternates = false
    private val inkBounds = Rect()

    init {
        installBackground()
        icon = defaultIcon()
    }

    fun apply(theme: KeyboardTheme) {
        this.theme = theme
        installBackground()
        invalidate()
    }

    private fun defaultIcon(): Drawable? = when (spec.cap) {
        KeyCap.Backspace -> ContextCompat.getDrawable(context, R.drawable.ic_backspace)
        KeyCap.Globe -> ContextCompat.getDrawable(context, R.drawable.ic_language)
        KeyCap.Shift -> ContextCompat.getDrawable(context, R.drawable.ic_shift)
        else -> null
    }

    private fun fontSizeSp(cap: KeyCap): Float = when (cap) {
        is KeyCap.Char, is KeyCap.InsertLiteral -> 22f
        KeyCap.Space, KeyCap.Return, is KeyCap.ToLayer, KeyCap.Globe, KeyCap.Spacer -> 16f
        KeyCap.Shift, KeyCap.Backspace -> 20f
    }

    private val isLightKey: Boolean get() = when (spec.cap) {
        is KeyCap.Char, is KeyCap.InsertLiteral, KeyCap.Space -> true
        else -> false
    }
    private val isReturnKey get() = spec.cap is KeyCap.Return

    private fun keyFill(): Int = when {
        isAccented -> theme.keyBackground       // active shift: lighter surface
        isReturnKey -> theme.returnBackground
        isLightKey -> theme.keyBackground
        else -> theme.specialKeyBackground
    }

    private fun textColor(): Int = when {
        isReturnKey -> theme.returnText
        isLightKey -> theme.keyText
        else -> theme.specialKeyText
    }

    // Rounded surface + a bounded Material ripple (clipped to the rounded mask).
    private fun installBackground() {
        val content = GradientDrawable().apply {
            shape = GradientDrawable.RECTANGLE
            cornerRadius = corner
            setColor(keyFill())
        }
        val mask = GradientDrawable().apply {
            shape = GradientDrawable.RECTANGLE
            cornerRadius = corner
            setColor(Color.WHITE)
        }
        background = RippleDrawable(ColorStateList.valueOf(theme.rippleColor), content, mask)
    }

    override fun onDraw(canvas: Canvas) {
        // Background (rounded surface + ripple) is drawn by View; we draw the icon or glyph.
        val ic = icon
        if (ic != null) {
            val size = (24 * density).toInt()
            val l = (width - size) / 2
            val t = (height - size) / 2
            ic.setBounds(l, t, l + size, t + size)
            ic.setTint(textColor())
            ic.draw(canvas)
            return
        }
        textPaint.color = textColor()
        val fm = textPaint.fontMetrics
        val baseline = height / 2f - (fm.ascent + fm.descent) / 2f
        var x = width / 2f
        if (spec.cap is KeyCap.InsertLiteral && displayText.isNotEmpty()) {
            // Fullwidth CJK punctuation (。，、！…) inks only the bottom-left of a
            // full-em advance. CoreText hides that on iOS by compressing the trailing
            // half-em at line ends; Android's Paint has no such rule, so centering the
            // advance leaves the mark visibly left of center. Center on the ink box
            // instead — horizontally only: the low vertical position is how these
            // marks sit in running text (and on the iOS keys).
            textPaint.getTextBounds(displayText, 0, displayText.length, inkBounds)
            x += textPaint.measureText(displayText) / 2f - inkBounds.exactCenterX()
        }
        canvas.drawText(displayText, x, baseline, textPaint)
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        when (event.actionMasked) {
            MotionEvent.ACTION_DOWN -> {
                isPressed = true
                drawableHotspotChanged(event.x, event.y)
                preview(KeyPreviewEvent.SHOW)
                if (spec.alternates.isNotEmpty()) startAlternatesDelay()
                if (isRepeating) { fire(); startRepeat() }
                return true
            }
            MotionEvent.ACTION_MOVE -> {
                if (showingAlternates) {
                    onAlternatesSlide?.invoke(this, event.x, event.y)
                    return true
                }
                val inside = insideExpanded(event.x, event.y)
                if (inside) {
                    if (!isPressed) { isPressed = true; preview(KeyPreviewEvent.SHOW) }
                    drawableHotspotChanged(event.x, event.y)
                } else {
                    if (isPressed) { isPressed = false; preview(KeyPreviewEvent.CANCEL) }
                    stopRepeat()
                    stopAlternatesDelay()
                }
                return true
            }
            MotionEvent.ACTION_UP -> {
                stopAlternatesDelay()
                if (showingAlternates) {
                    showingAlternates = false
                    isPressed = false
                    val value = onAlternatesCommit?.invoke(this)
                    if (value != null) {
                        // The primary commits like a plain tap (a future Char
                        // group must still go through the engine); alternates
                        // bypass it.
                        if (value == displayText) fire() else onTap?.invoke(KeyCap.InsertLiteral(value))
                    }
                    return true
                }
                val inside = insideExpanded(event.x, event.y)
                isPressed = false
                stopRepeat()
                preview(KeyPreviewEvent.RELEASE)
                if (!isRepeating && inside) fire()
                return true
            }
            MotionEvent.ACTION_CANCEL -> {
                stopAlternatesDelay()
                showingAlternates = false
                isPressed = false
                stopRepeat()
                preview(KeyPreviewEvent.CANCEL)
                return true
            }
        }
        return super.onTouchEvent(event)
    }

    private fun insideExpanded(x: Float, y: Float): Boolean =
        x >= -expand && y >= -expand && x <= width + expand && y <= height + expand

    private fun fire() = onTap?.invoke(spec.cap)

    private fun preview(event: KeyPreviewEvent) {
        if (hasPreview) onPreview?.invoke(this, event)
    }

    // Fire immediately on down, then after 350ms repeat every 100ms (matches iOS).
    private fun startRepeat() {
        val r = object : Runnable {
            override fun run() { fire(); handler.postDelayed(this, 100) }
        }
        repeatRunnable = r
        handler.postDelayed(r, 350)
    }

    private fun stopRepeat() {
        repeatRunnable?.let { handler.removeCallbacks(it) }
        repeatRunnable = null
    }

    // Hold ~450ms on a grouped key to swap the balloon for the alternates row.
    private fun startAlternatesDelay() {
        val r = Runnable {
            alternatesRunnable = null
            if (onAlternatesExpand?.invoke(this) == true) showingAlternates = true
        }
        alternatesRunnable = r
        handler.postDelayed(r, 450)
    }

    private fun stopAlternatesDelay() {
        alternatesRunnable?.let { handler.removeCallbacks(it) }
        alternatesRunnable = null
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        stopRepeat()
        stopAlternatesDelay()
        showingAlternates = false
        // Layer switches rebuild the key plane mid-press; don't strand a balloon.
        preview(KeyPreviewEvent.CANCEL)
    }
}
